### PR TITLE
chore: add wunderstack config rfc

### PR DESCRIPTION
**Pull request by Pagebakers**
#### Checklist

- [ ] run `make test`
- [ ] PR title must follow [conventional-commit-standard](https://github.com/wundergraph/wundergraph/blob/main/CONTRIBUTING.md#conventional-commit-standard)
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Code of conduct](https://github.com/wundergraph/wundergraph/blob/main/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by YuriBuerov** [comment]
The idea of the stack was - the user does not need to configure anything, we provide it out of the box. The only case when you need to configure storage/kV is when you need a custom solution, or if you want to disable it (tbd)

**Comment by Pagebakers** [comment]
> The idea of the stack was - the user does not need to configure anything, we provide it out of the box. The only case when you need to configure storage/kV is when you need a custom solution, or if you want to disable it (tbd)

Yep that is described in the RFC. You only add the stack that you need, without configuring credentials, that is automatically provisioned. You still need to configure a name, bucketName and uploadProfiles in case of the s3 integration.

**Review by JivusAyrus — COMMENTED** [review]
Looks good

**Review by jensneuse — COMMENTED** [review]
I've added a few comments.

One important thing I think is missing in this RFC is that we've got to think about multiple things:

How can a stack be integrated into the context object. We're providing the context/ctx object to all hooks and handlers. How can we make it easy for the user to "extend" the context with their own stack thing. E.g. when you add an S3 storage or a KV, how can the creator of a stack extend the context object.

Next, what should happen during deployment time? Some stacks, like a Neon/Planetscale/Cockraoch Database will require migration steps when a stack gets deployed. The creator of a stack needs to be able to hook into the deployment lifecycle to manage the stack migrations, etc... We should come up with a solution to allow anybody to create a stack that works both locally and also in the cloud.

Finally, a stack needs to work in local development. This means, the creator of a stack, which could also be us, needs to be able to define the behavior of the stack during local development. When the user runs `wunderctl up` how can the stack define what to do? (lifecycle events). When the user stops `wunderctl up` what should happen? (cleanup)

When the user runs generate in the CI, should we run some migration step? Or should this step run on the first start of the stack during deployment?

How will a stack behave in a preview environment? What happens when a new branch gets deployed? What happens when the branch is deleted?

I think the solution to stack configuration needs to be generic enough to support all these use cases. We can start with simpler stacks that allow us to implement e.g. KV, but we should already aim for a generic solution.

Ideally, anybody, like a database provider company, can create their own stack which hooks into the lifecycle events of the WunderGraph application. We can offer them an API to start and stop docker containers locally, while other integration points will allow them to configure, maintain and run their stacks alongside a WunderGraph application in the cloud.

Another question is, when you define a stack, can the user get information on the state? E.g. there could be another API to allow the stack creator to expose an interface that can be used to understand the state of the integration. E.g. by implementing the interface, the user could see or configure their stack in the cloud UI. Let's say you're defining a secret for a database in the stack. When you deploy this to cloud, you'll get an info in the stacks tab that you've got to login or set a secret. Once the connection is established, the stack can show it's state.

File: rfc/wunderstack-config/[RFC]wunderstack-config-001-Pagebakers.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+### KV
+
+Key-value stores are configured with the `kv` property. It works similar to storage and authentication. `WunderKv` is managed by WunderGraph cloud, configuration is injected into the config at build time. There can be other providers like a generic Redis provider `kvProviders.redis`.
+
+```ts
+import { configureWunderGraphApplication, kvProviders } from '@wundergraph/sdk';
+
+configureWunderGraphApplication({
+  apis: [],
+  kv: [
```
[the remark this answers is not in the document]
Comment (StarpTech) [reply rfc/wunderstack-config/[RFC]wunderstack-config-001-Pagebakers.md:87]: I agree. We can identify the resource type based on the property of a class or object.

**Comment by Pagebakers** [comment]
> I've added a few comments.
> 
> One important thing I think is missing in this RFC is that we've got to think about multiple things:
> 

This RFC doesn't propose a full solution for supporting stacks, but it's more a solution that will work with our existing APIs in the short term and still allows us to define 'WunderGraph' primitives that auto configure s3/kv and deploy it to WunderGraph cloud, which we want to release in 2 weeks. It allows us to start testing the internal 'stacks' technology without introducing new APIs and above all start getting feedback from devs. While we work out all the details for the complete stacks spec, it's a complicated problem to solve right, and definitely we need to take the time to design it properly.

File: rfc/wunderstack-config/[RFC]wunderstack-config-001-Pagebakers.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Pagebakers) [reply rfc/wunderstack-config/[RFC]wunderstack-config-001-Pagebakers.md:87]: Perhaps it won't scale, but the goal is to release something that works by the end of this month. We can experiment with WunderGraph primitives already by using our existing config API, just like the auth config `authentication.auth0()`  vs `s3Storage.wundergraph()`.

I see the WunderGraph config as a more low level implementation, and stacks would sit on top of it, abstracting away the configuration/life-cycles/events/etc. Perhaps similar to what CDK is to cloud formation.

**Pull request closed without merging**